File: sdk/javascript/packages/core/test/record_link.test.ts
import {KeeperRecordLink, getLinks, KeeperRecord, platform} from '../'
import {createCipheriv, randomBytes} from 'crypto'

// KSM-1010: KeeperRecordLink typed accessor tests (mirrors Python record_link_test.py)

const plainLink = (payload: object, path?: string, ownerRecordUid = 'RU_owner'): KeeperRecordLink => {
    const data = platform.bytesToBase64(platform.stringToBytes(JSON.stringify(payload)))
    return new KeeperRecordLink({recordUid: 'RU_test', data, path}, ownerRecordUid)
}

const encryptedLink = async (payload: object, key: Uint8Array, path?: string, ownerRecordUid = 'RU_owner'): Promise<KeeperRecordLink> => {
    const encrypted = await platform.encryptWithKey(platform.stringToBytes(JSON.stringify(payload)), key)
    const data = platform.bytesToBase64(encrypted)
    return new KeeperRecordLink({recordUid: 'RU_test', data, path}, ownerRecordUid)
}

// Encrypt with a specific IV so we can control the first decoded byte (for the fallthrough test)
const encryptWithCustomIv = (data: Uint8Array, key: Uint8Array, iv: Uint8Array): Uint8Array => {
    const cipher = createCipheriv('aes-256-gcm', Buffer.from(key), Buffer.from(iv)) as any
    const ct: Buffer = Buffer.concat([cipher.update(Buffer.from(data)), cipher.final()])
    const tag: Buffer = cipher.getAuthTag()
    const result = new Uint8Array(iv.length + ct.length + tag.length)
    result.set(iv, 0)
    result.set(ct, iv.length)
    result.set(tag, iv.length + ct.length)
    return result
}

// ── test 1 ─────────────────────────────────────────────────────────────────────
test('boolean accessors read plain JSON; absent keys default to false', () => {
    const link = plainLink({is_admin: true, rotation: true, connections: false})
    expect(link.isAdminUser()).toBe(true)
    expect(link.allowsRotation()).toBe(true)
    expect(link.allowsConnections()).toBe(false)
    expect(link.allowsPortForwards()).toBe(false)
    expect(link.isLaunchCredential()).toBe(false)
    expect(link.isIamUser()).toBe(false)
    expect(link.belongsTo()).toBe(false)
    expect(link.noUpdateServices()).toBe(false)
})

// ── test 2 ─────────────────────────────────────────────────────────────────────
test('version, decoded data, and readable-JSON heuristic', () => {
    const link = plainLink({version: 3, is_admin: false})
    expect(link.getLinkDataVersion()).toBe(3)
    const decoded = link.getDecodedData()
    expect(decoded).not.toBeNull()
    expect(decoded!.startsWith('{')).toBe(true)
    expect(link.hasReadableData()).toBe(true)

    // Non-JSON base64 is not "readable"
    const raw = new KeeperRecordLink({
        recordUid: 'RU',
        data: platform.bytesToBase64(platform.stringToBytes('not json at all')),
        path: undefined
    }, 'owner')
    expect(raw.hasReadableData()).toBe(false)
    expect(raw.getLinkDataVersion()).toBeNull()

    // Invalid base64 returns null (Node's Buffer.from is lenient; we validate before decoding)
    const bad = new KeeperRecordLink({recordUid: 'RU', data: '!!! not base64 !!!', path: undefined}, 'owner')
    expect(bad.getDecodedData()).toBeNull()
    expect(() => bad.getDecodedData()).not.toThrow()
})

// ── test 3 ─────────────────────────────────────────────────────────────────────
test('mightBeEncrypted is gated to known encrypted paths only', () => {
    expect(plainLink({}, 'ai_settings').mightBeEncrypted()).toBe(true)
    expect(plainLink({}, 'jit_settings').mightBeEncrypted()).toBe(true)
    expect(plainLink({}, 'meta').mightBeEncrypted()).toBe(false)
    expect(plainLink({}, 'something_else').mightBeEncrypted()).toBe(false)
    expect(plainLink({}, undefined).mightBeEncrypted()).toBe(false)
})

// ── test 4 ─────────────────────────────────────────────────────────────────────
test('getDecryptedData roundtrip: correct key decrypts, wrong/absent key yields null', async () => {
    const key = platform.getRandomBytes(32)
    const payload = {enabled: true, ttl: 3600}
    const link = await encryptedLink(payload, key, 'jit_settings')

    const decrypted = await link.getDecryptedData(key)
    expect(decrypted).not.toBeNull()
    expect(JSON.parse(decrypted!)).toEqual(payload)

    expect(await link.getDecryptedData(undefined)).toBeNull()

    const wrongKey = platform.getRandomBytes(32)
    expect(await link.getDecryptedData(wrongKey)).toBeNull()
})

// ── test 5 ─────────────────────────────────────────────────────────────────────
test('getLinkData auto-detects plain JSON vs encrypted data', async () => {
    const plain = plainLink({aiEnabled: true}, 'ai_settings')
    const data = await plain.getLinkData()
    expect(data).not.toBeNull()
    expect(data!['aiEnabled']).toBe(true)

    const key = platform.getRandomBytes(32)
    const enc = await encryptedLink({enabled: true}, key, 'jit_settings')
    expect(await enc.getLinkData()).toBeNull()
    const withKey = await enc.getLinkData(key)
    expect(withKey).not.toBeNull()
    expect(withKey!['enabled']).toBe(true)
})

// ── test 6 ─────────────────────────────────────────────────────────────────────
test('settings accessors are gated to the matching path', async () => {
    const key = platform.getRandomBytes(32)
    const ai = plainLink({aiEnabled: true}, 'ai_settings')
    const jit = plainLink({enabled: true}, 'jit_settings')

    expect(await ai.getAiSettingsData(key)).not.toBeNull()
    expect(await ai.getJitSettingsData(key)).toBeNull()
    expect(await jit.getJitSettingsData(key)).not.toBeNull()
    expect(await jit.getAiSettingsData(key)).toBeNull()

    expect(await ai.getSettingsForPath('ai_settings')).not.toBeNull()
    expect(await ai.getSettingsForPath('other')).toBeNull()
})

// ── test 7 ─────────────────────────────────────────────────────────────────────
test('getLinks wraps typed links; entry without recordUid is skipped', () => {
    const record: KeeperRecord = {
        recordUid: 'main_uid',
        data: {},
        revision: 1,
        links: [
            {recordUid: 'meta_uid', data: platform.bytesToBase64(platform.stringToBytes(
                JSON.stringify({allowedSettings: {rotation: true}, version: 1}))), path: 'meta'},
            {recordUid: 'cred_uid', data: platform.bytesToBase64(platform.stringToBytes(
                JSON.stringify({is_admin: true, is_launch_credential: true}))), path: undefined},
            {recordUid: 'ref_uid', data: undefined, path: undefined},
            {recordUid: '', data: undefined, path: undefined}, // empty UID is skipped
        ]
    }

    const links = getLinks(record)
    expect(links).toHaveLength(3)
    expect(links.every(l => l instanceof KeeperRecordLink)).toBe(true)

    expect(links[0].recordUid).toBe('meta_uid')
    expect(links[0].path).toBe('meta')
    expect(links[0].allowsRotation()).toBe(true)
    expect(links[0].getLinkDataVersion()).toBe(1)

    expect(links[1].isAdminUser()).toBe(true)
    expect(links[1].isLaunchCredential()).toBe(true)

    expect(links[2].recordUid).toBe('ref_uid')

    // Raw links unchanged
    expect(record.links).toHaveLength(4)
})

// ── test 8 ─────────────────────────────────────────────────────────────────────
test('string-encoded values are not coerced to bool or int', () => {
    const link = plainLink({is_admin: 'true', rotation: 'false', version: '3'})
    expect(link.isAdminUser()).toBe(false)
    expect(link.allowsRotation()).toBe(false)
    expect(link.getLinkDataVersion()).toBeNull()

    const typed = plainLink({is_admin: true, version: 3})
    expect(typed.isAdminUser()).toBe(true)
    expect(typed.getLinkDataVersion()).toBe(3)

    // Boolean is not a valid integer version
    const boolVersion = plainLink({version: true})
    expect(boolVersion.getLinkDataVersion()).toBeNull()
})

// ── test 9 ─────────────────────────────────────────────────────────────────────
test('hasEncryptedData detects ciphertext; printable text and JSON are not encrypted', async () => {
    const key = platform.getRandomBytes(32)
    const enc = await encryptedLink({secret: 'value'}, key)
    expect(enc.hasEncryptedData()).toBe(true)

    const text = new KeeperRecordLink({
        recordUid: 'RU',
        data: platform.bytesToBase64(platform.stringToBytes('just plain readable text, not json')),
        path: undefined
    }, 'owner')
    expect(text.hasEncryptedData()).toBe(false)

    expect(plainLink({a: 1}).hasEncryptedData()).toBe(false)

    const noData = new KeeperRecordLink({recordUid: 'RU', data: undefined, path: undefined}, 'owner')
    expect(noData.hasEncryptedData()).toBe(false)
})

// ── test 10 ────────────────────────────────────────────────────────────────────
test('getSettingsForPath decrypts encrypted payload for a matching path', async () => {
    const key = platform.getRandomBytes(32)
    const link = await encryptedLink({customSetting: 42}, key, 'custom_settings')

    const data = await link.getSettingsForPath('custom_settings', key)
    expect(data).not.toBeNull()
    expect(data!['customSetting']).toBe(42)
    expect(await link.getSettingsForPath('other', key)).toBeNull()
})

// ── test 11 ────────────────────────────────────────────────────────────────────
test('meta link: permission booleans fall back to allowedSettings', async () => {
    const link = plainLink({
        allowedSettings: {
            rotation: true, connections: true, portForwards: true,
            sessionRecording: true, typescriptRecording: false,
            aiEnabled: true, aiSessionTerminate: true, remoteBrowserIsolation: true
        },
        rotateOnTermination: false,
        version: 1,
        no_update_services: true
    }, 'meta')

    expect(link.allowsRotation()).toBe(true)
    expect(link.allowsConnections()).toBe(true)
    expect(link.allowsPortForwards()).toBe(true)
    expect(link.allowsSessionRecording()).toBe(true)
    expect(link.allowsTypescriptRecording()).toBe(false)
    expect(link.allowsRemoteBrowserIsolation()).toBe(true)
    expect(link.aiEnabled()).toBe(true)
    expect(link.aiSessionTerminate()).toBe(true)

    expect(link.rotatesOnTermination()).toBe(false)
    expect(link.getLinkDataVersion()).toBe(1)
    expect(link.noUpdateServices()).toBe(true)

    const allowed = link.getAllowedSettings()
    expect(allowed['rotation']).toBe(true)

    const meta = await link.getMetaData()
    expect(meta).not.toBeNull()
    expect(meta!['version']).toBe(1)
    expect(await plainLink({}, undefined).getMetaData()).toBeNull()
})

// ── test 12 ────────────────────────────────────────────────────────────────────
test('credential link: user flags and rotation_settings object', () => {
    const link = plainLink({
        is_admin: true, is_iam_user: false, belongs_to: true, is_launch_credential: true,
        rotation_settings: {schedule: '', pwd_complexity: 'ZmFrZS1jb21wbGV4aXR5',
            disabled: false, noop: false, saas_record_uid_list: []}
    })

    expect(link.isAdminUser()).toBe(true)
    expect(link.isIamUser()).toBe(false)
    expect(link.belongsTo()).toBe(true)
    expect(link.isLaunchCredential()).toBe(true)

    const rs = link.getRotationSettings()
    expect(rs).not.toBeNull()
    expect(rs!['schedule']).toBe('')
    expect(rs!['disabled']).toBe(false)
    expect(rs!['saas_record_uid_list']).toEqual([])

    expect(plainLink({is_admin: true}).getRotationSettings()).toBeNull()
})

// ── test 13 ────────────────────────────────────────────────────────────────────
test('data-less reference link answers all accessors with false/null', async () => {
    const link = new KeeperRecordLink({recordUid: 'RU_ref', data: undefined, path: undefined}, 'owner')

    expect(link.recordUid).toBe('RU_ref')
    expect(link.isAdminUser()).toBe(false)
    expect(link.allowsRotation()).toBe(false)
    expect(link.getLinkDataVersion()).toBeNull()
    expect(link.getDecodedData()).toBeNull()
    expect(await link.getDecryptedData(platform.getRandomBytes(32))).toBeNull()
    expect(await link.getLinkData()).toBeNull()
    expect(link.getAllowedSettings()).toEqual({})
    expect(link.getRotationSettings()).toBeNull()
    expect(link.hasReadableData()).toBe(false)
    expect(link.hasEncryptedData()).toBe(false)
})

// ── test 14 ────────────────────────────────────────────────────────────────────
test('ai_settings link decrypts to riskLevels payload', async () => {
    const key = platform.getRandomBytes(32)
    const payload = {
        version: 'v1.0.0',
        riskLevels: {
            critical: {tags: {allow: [], deny: []}, aiSessionTerminate: true},
            high: {tags: {allow: [], deny: []}, aiSessionTerminate: true},
            low: {tags: {allow: []}, aiSessionTerminate: false}
        }
    }
    const link = await encryptedLink(payload, key, 'ai_settings')

    const data = await link.getAiSettingsData(key)
    expect(data).toEqual(payload)
    expect(link.getLinkDataVersion()).toBeNull() // version is a string here
})

// ── test 15 ────────────────────────────────────────────────────────────────────
test('jit_settings link decrypts to elevation payload', async () => {
    const key = platform.getRandomBytes(32)
    const payload = {
        createEphemeral: true, elevate: true,
        elevationMethod: 'group', elevationString: 'arn:aws', baseDistinguishedName: ''
    }
    const link = await encryptedLink(payload, key, 'jit_settings')

    expect(await link.getJitSettingsData(key)).toEqual(payload)
})

// ── test 16 ────────────────────────────────────────────────────────────────────
test('top-level boolean wins over allowedSettings fallback', () => {
    const link = plainLink({rotation: false, allowedSettings: {rotation: true}})
    expect(link.allowsRotation()).toBe(false)

    const onlyNested = plainLink({allowedSettings: {rotation: true}})
    expect(onlyNested.allowsRotation()).toBe(true)
})

// ── test 17 ────────────────────────────────────────────────────────────────────
test('ciphertext coincidentally starting with { or [ still decrypts (fallthrough)', async () => {
    const key = platform.getRandomBytes(32)
    const payload = {createEphemeral: true, elevate: true}
    const plaintext = platform.stringToBytes(JSON.stringify(payload))

    for (const marker of [0x7b /* { */, 0x5b /* [ */]) {
        const iv = new Uint8Array(12)
        iv[0] = marker
        randomBytes(11).copy(Buffer.from(iv.buffer), 1)

        const ciphertext = encryptWithCustomIv(plaintext, key, iv)
        const link = new KeeperRecordLink({
            recordUid: 'RU',
            data: platform.bytesToBase64(ciphertext),
            path: 'jit_settings'
        }, 'owner')

        const decoded = link.getDecodedData()
        expect(decoded).not.toBeNull()
        expect(decoded!.charCodeAt(0)).toBe(marker)

        expect(await link.getLinkData(key)).toEqual(payload)
        expect(await link.getJitSettingsData(key)).toEqual(payload)
        expect(await link.getSettingsForPath('jit_settings', key)).toEqual(payload)
        expect(await link.getLinkData(undefined)).toBeNull()
    }

    // Plain JSON fast path is unaffected
    expect(await plainLink({a: 1}).getLinkData()).toEqual({a: 1})
})

File: sdk/python/core/tests/record_link_test.py
import json
import logging
import os
import tempfile
import unittest

from keeper_secrets_manager_core import SecretsManager
from keeper_secrets_manager_core import mock, utils
from keeper_secrets_manager_core.crypto import CryptoUtils
from keeper_secrets_manager_core.dto.dtos import KeeperRecordLink
from keeper_secrets_manager_core.mock import MockConfig
from keeper_secrets_manager_core.storage import FileKeyValueStorage


def plain_link(payload, path=None, record_uid="RU_test"):
    """Build a link whose data is base64 of the given plain JSON payload."""
    return KeeperRecordLink({
        "recordUid": record_uid,
        "data": utils.bytes_to_base64(json.dumps(payload).encode()),
        "path": path
    })


def encrypted_link(payload, key, path=None, record_uid="RU_test"):
    """Build a link whose data is base64 of the payload encrypted with AES-256-GCM."""
    ciphertext = CryptoUtils.encrypt_aes(json.dumps(payload).encode(), key)
    return KeeperRecordLink({
        "recordUid": record_uid,
        "data": utils.bytes_to_base64(ciphertext),
        "path": path
    })


class RecordLinkTest(unittest.TestCase):

    def setUp(self):

        self.orig_working_dir = os.getcwd()

        logger = logging.getLogger("ksm")
        logger.setLevel(logging.DEBUG)
        logger.propagate = False
        while logger.hasHandlers():
            logger.removeHandler(logger.handlers[0])
        handler = logging.StreamHandler()
        logger.addHandler(handler)
        formatter = logging.Formatter(f'%(asctime)s %(name)s  %(levelname)s: %(message)s')
        handler.setFormatter(formatter)

    def tearDown(self):

        os.chdir(self.orig_working_dir)

    def test_boolean_accessors_read_plain_json(self):
        """Boolean accessors read the decoded plain JSON; absent keys default to False."""

        link = plain_link({"is_admin": True, "rotation": True, "connections": False})

        self.assertTrue(link.is_admin_user(), "is_admin true should read true")
        self.assertTrue(link.allows_rotation(), "rotation true should read true")
        self.assertFalse(link.allows_connections(), "connections false should read false")

        # Absent keys must default to False
        self.assertFalse(link.allows_port_forwards(), "absent key must default to False")
        self.assertFalse(link.is_launch_credential(), "absent key must default to False")
        self.assertFalse(link.is_iam_user(), "absent key must default to False")
        self.assertFalse(link.belongs_to(), "absent key must default to False")
        self.assertFalse(link.no_update_services(), "absent key must default to False")

    def test_version_and_decoded_data(self):
        """Integer version, decoded data and the readable-JSON heuristic."""

        link = plain_link({"version": 3, "is_admin": False})
        self.assertEqual(3, link.get_link_data_version())
        decoded = link.get_decoded_data()
        self.assertIsNotNone(decoded, "plain JSON data should decode")
        self.assertTrue(decoded.startswith("{"), "decoded data is the raw JSON")
        self.assertTrue(link.has_readable_data(), "JSON payload is readable")

        # Non-JSON (but valid base64) decoded content is not "readable"
        raw = KeeperRecordLink({
            "recordUid": "RU",
            "data": utils.bytes_to_base64(b"not json at all"),
            "path": None
        })
        self.assertFalse(raw.has_readable_data(), "plain text without JSON markers is not readable")
        self.assertIsNone(raw.get_link_data_version(), "no version in non-JSON data")

        # Invalid base64 decodes to None, never raises
        bad = KeeperRecordLink({"recordUid": "RU", "data": "!!! not base64 !!!", "path": None})
        self.assertIsNone(bad.get_decoded_data())
        self.assertIsNone(bad.get_link_data())

    def test_might_be_encrypted_by_path(self):
        """might_be_encrypted is gated to the known encrypted paths only."""

        self.assertTrue(plain_link({}, path="ai_settings").might_be_encrypted())
        self.assertTrue(plain_link({}, path="jit_settings").might_be_encrypted())
        self.assertFalse(plain_link({}, path="meta").might_be_encrypted(),
                         "meta links carry plain JSON")
        self.assertFalse(plain_link({}, path="something_else").might_be_encrypted(),
                         "unknown path must not be assumed encrypted")
        self.assertFalse(plain_link({}, path=None).might_be_encrypted())

    def test_get_decrypted_data_roundtrip(self):
        """AES-256-GCM decrypt round-trip with the record key; wrong/absent key gives None."""

        key = CryptoUtils.generate_random_bytes(32)
        payload = {"enabled": True, "ttl": 3600}
        link = encrypted_link(payload, key, path="jit_settings")

        decrypted = link.get_decrypted_data(key)
        self.assertIsNotNone(decrypted, "correct key decrypts")
        self.assertEqual(payload, json.loads(decrypted), "decrypts to the original plaintext")

        self.assertIsNone(link.get_decrypted_data(None), "no key gives None")

        wrong_key = CryptoUtils.generate_random_bytes(32)
        self.assertIsNone(link.get_decrypted_data(wrong_key),
                          "wrong key fails to decrypt and gives None, not an exception")

    def test_get_link_data_plain_and_encrypted(self):
        """get_link_data auto-detects plain JSON vs encrypted data."""

        plain = plain_link({"aiEnabled": True}, path="ai_settings")
        data = plain.get_link_data()
        self.assertIsNotNone(data, "plain JSON parses without a key")
        self.assertTrue(data.get("aiEnabled"))

        key = CryptoUtils.generate_random_bytes(32)
        enc = encrypted_link({"enabled": True}, key, path="jit_settings")
        self.assertIsNone(enc.get_link_data(), "encrypted without key gives None")
        data = enc.get_link_data(key)
        self.assertIsNotNone(data, "encrypted with key parses")
        self.assertTrue(data.get("enabled"))

    def test_settings_path_filters(self):
        """Settings accessors are gated to the matching path."""

        key = CryptoUtils.generate_random_bytes(32)
        ai = plain_link({"aiEnabled": True}, path="ai_settings")
        jit = plain_link({"enabled": True}, path="jit_settings")

        self.assertIsNotNone(ai.get_ai_settings_data(key), "ai path returns ai settings")
        self.assertIsNone(ai.get_jit_settings_data(key), "ai path is not jit")
        self.assertIsNotNone(jit.get_jit_settings_data(key), "jit path returns jit settings")
        self.assertIsNone(jit.get_ai_settings_data(key), "jit path is not ai")

        # Generic accessor matches any path
        self.assertIsNotNone(ai.get_settings_for_path("ai_settings"),
                             "generic accessor matches the path")
        self.assertIsNone(ai.get_settings_for_path("other"),
                          "generic accessor returns None for a non-matching path")

    def test_record_get_links_end_to_end(self):
        """Record.get_links() builds typed links from the raw links field, which is unchanged."""

        try:
            with tempfile.NamedTemporaryFile("w", delete=False) as fh:
                fh.write(MockConfig.make_json())
                fh.seek(0)
                secrets_manager = SecretsManager(
                    config=FileKeyValueStorage(config_file_location=fh.name)
                )

                res = mock.Response()

                linked_record = res.add_record(title="Linked User", record_type='login')
                linked_record.field("login", "linkeduser")

                links_data = [
                    # meta self-link (plain JSON, live shape)
                    {"recordUid": "mainUid", "path": "meta",
                     "data": utils.bytes_to_base64(json.dumps(
                         {"allowedSettings": {"rotation": True}, "version": 1}).encode())},
                    # credential link to another record
                    {"recordUid": linked_record.uid, "path": None,
                     "data": utils.bytes_to_base64(json.dumps(
                         {"is_admin": True, "is_launch_credential": True}).encode())},
                    # pure reference link (no data)
                    {"recordUid": "referencedUid", "data": None, "path": None},
                    # malformed entry without recordUid is kept raw but skipped by get_links()
                    {"data": None, "path": None},
                ]
                main_record = res.add_record(
                    title="Main Record",
                    record_type='login',
                    links=links_data
                )
                main_record.field("login", "mainuser")
                main_record.field("password", "mainpass")

                res_queue = mock.ResponseQueue(client=secrets_manager)
                res_queue.add_response(res)

                records = secrets_manager.get_secrets()

                main_rec = next((r for r in records if r.title == "Main Record"), None)
                self.assertIsNotNone(main_rec, "Main record not found")

                # The raw links field is unchanged (back-compat)
                self.assertEqual(4, len(main_rec.links), "raw links list keeps all entries")
                self.assertEqual(links_data, main_rec.links, "raw links are untouched dicts")

                links = main_rec.get_links()
                self.assertEqual(3, len(links), "entry without recordUid is skipped")
                self.assertTrue(all(isinstance(link, KeeperRecordLink) for link in links))

                meta = links[0]
                self.assertEqual("mainUid", meta.record_uid)
                self.assertEqual("meta", meta.path)
                self.assertTrue(meta.allows_rotation(), "typed link decodes its data")
                self.assertEqual(1, meta.get_link_data_version())

                cred = links[1]
                self.assertEqual(linked_record.uid, cred.record_uid)
                self.assertTrue(cred.is_admin_user())
                self.assertTrue(cred.is_launch_credential())

                ref = links[2]
                self.assertEqual("referencedUid", ref.record_uid)
                self.assertIsNone(ref.get_link_data())
        finally:
            try:
                os.unlink(fh.name)
            except OSError:
                pass

    def test_string_encoded_values_are_not_coerced(self):
        """String-encoded values are NOT coerced to bool/int (Java/Rust parity)."""

        link = plain_link({"is_admin": "true", "rotation": "false", "version": "3"})
        self.assertFalse(link.is_admin_user(), 'string "true" is not coerced to bool')
        self.assertFalse(link.allows_rotation(), 'string "false" stays falsey')
        self.assertIsNone(link.get_link_data_version(), 'string "3" is not coerced to int')

        # Real JSON bool/number ARE read
        typed = plain_link({"is_admin": True, "version": 3})
        self.assertTrue(typed.is_admin_user())
        self.assertEqual(3, typed.get_link_data_version())

        # Python quirk: True is an int, but must not count as a version
        bool_version = plain_link({"version": True})
        self.assertIsNone(bool_version.get_link_data_version(),
                          "boolean version is not an integer version")

    def test_has_encrypted_data_detection(self):
        """has_encrypted_data inspects the content: ciphertext yes, printable text/JSON no."""

        key = CryptoUtils.generate_random_bytes(32)
        ciphertext = CryptoUtils.encrypt_aes(b"some secret bytes", key)
        enc = KeeperRecordLink({
            "recordUid": "RU",
            "data": utils.bytes_to_base64(ciphertext),
            "path": None
        })
        self.assertTrue(enc.has_encrypted_data(), "ciphertext is detected as encrypted")

        text = KeeperRecordLink({
            "recordUid": "RU",
            "data": utils.bytes_to_base64(b"just plain readable text, not json"),
            "path": None
        })
        self.assertFalse(text.has_encrypted_data(), "printable text is not flagged encrypted")

        self.assertFalse(plain_link({"a": 1}).has_encrypted_data(),
                         "JSON is not flagged encrypted")

        no_data = KeeperRecordLink({"recordUid": "RU", "data": None, "path": None})
        self.assertFalse(no_data.has_encrypted_data(), "no data is not flagged encrypted")

    def test_get_settings_for_path_encrypted(self):
        """Generic settings accessor decrypts an encrypted payload for a matching path."""

        key = CryptoUtils.generate_random_bytes(32)
        link = encrypted_link({"customSetting": 42}, key, path="custom_settings")

        data = link.get_settings_for_path("custom_settings", key)
        self.assertIsNotNone(data, "matching path with key decrypts")
        self.assertEqual(42, data.get("customSetting"))
        self.assertIsNone(link.get_settings_for_path("other", key), "non-matching path gives None")

    def test_meta_link_live_shape(self):
        """meta self-links: permission booleans fall back to the nested allowedSettings."""

        link = plain_link({
            "allowedSettings": {
                "rotation": True,
                "connections": True,
                "portForwards": True,
                "sessionRecording": True,
                "typescriptRecording": False,
                "aiEnabled": True,
                "aiSessionTerminate": True,
                "remoteBrowserIsolation": True
            },
            "rotateOnTermination": False,
            "version": 1,
            "no_update_services": True
        }, path="meta")

        # Permission booleans read from allowedSettings when absent at the top level
        self.assertTrue(link.allows_rotation())
        self.assertTrue(link.allows_connections())
        self.assertTrue(link.allows_port_forwards())
        self.assertTrue(link.allows_session_recording())
        self.assertFalse(link.allows_typescript_recording())
        self.assertTrue(link.allows_remote_browser_isolation())
        self.assertTrue(link.ai_enabled())
        self.assertTrue(link.ai_session_terminate())

        # Top-level fields
        self.assertFalse(link.rotates_on_termination())
        self.assertEqual(1, link.get_link_data_version())
        self.assertTrue(link.no_update_services())

        # Dict accessors
        allowed = link.get_allowed_settings()
        self.assertTrue(allowed.get("rotation"))
        meta = link.get_meta_data()
        self.assertIsNotNone(meta, "meta data parses without a key")
        self.assertEqual(1, meta.get("version"))
        self.assertIsNone(plain_link({}, path=None).get_meta_data(),
                          "get_meta_data is gated to path meta")

    def test_credential_link_live_shape(self):
        """Credential links: user flags and the nested rotation_settings object."""

        link = plain_link({
            "is_admin": True,
            "is_iam_user": False,
            "belongs_to": True,
            "is_launch_credential": True,
            "rotation_settings": {
                "schedule": "",
                "pwd_complexity": "ZmFrZS1jb21wbGV4aXR5",
                "disabled": False,
                "noop": False,
                "saas_record_uid_list": []
            }
        })

        self.assertTrue(link.is_admin_user())
        self.assertFalse(link.is_iam_user())
        self.assertTrue(link.belongs_to())
        self.assertTrue(link.is_launch_credential())

        rotation_settings = link.get_rotation_settings()
        self.assertIsNotNone(rotation_settings)
        self.assertEqual("", rotation_settings.get("schedule"))
        self.assertFalse(rotation_settings.get("disabled"))
        self.assertEqual([], rotation_settings.get("saas_record_uid_list"))

        self.assertIsNone(plain_link({"is_admin": True}).get_rotation_settings(),
                          "absent rotation_settings gives None")

    def test_data_less_reference_link(self):
        """Pure reference links (data null) answer all accessors with False/None."""

        link = KeeperRecordLink({"recordUid": "RU_ref", "data": None, "path": None})

        self.assertEqual("RU_ref", link.record_uid)
        self.assertFalse(link.is_admin_user())
        self.assertFalse(link.allows_rotation())
        self.assertIsNone(link.get_link_data_version())
        self.assertIsNone(link.get_decoded_data())
        self.assertIsNone(link.get_decrypted_data(CryptoUtils.generate_random_bytes(32)))
        self.assertIsNone(link.get_link_data())
        self.assertEqual({}, link.get_allowed_settings())
        self.assertIsNone(link.get_rotation_settings())
        self.assertFalse(link.has_readable_data())
        self.assertFalse(link.has_encrypted_data())

    def test_ai_settings_live_shape(self):
        """ai_settings links decrypt to the current riskLevels payload."""

        key = CryptoUtils.generate_random_bytes(32)
        payload = {
            "version": "v1.0.0",
            "riskLevels": {
                "critical": {"tags": {"allow": [], "deny": []}, "aiSessionTerminate": True},
                "high": {"tags": {"allow": [], "deny": []}, "aiSessionTerminate": True},
                "medium": {"tags": {"allow": [], "deny": []}, "aiSessionTerminate": True},
                "low": {"tags": {"allow": []}, "aiSessionTerminate": False}
            }
        }
        link = encrypted_link(payload, key, path="ai_settings")

        data = link.get_ai_settings_data(key)
        self.assertIsNotNone(data)
        self.assertEqual(payload, data, "nested riskLevels structure is preserved")

        # The live version field is a string here, so the integer accessor yields None
        self.assertIsNone(link.get_link_data_version())

    def test_jit_settings_live_shape(self):
        """jit_settings links decrypt to the current elevation payload."""

        key = CryptoUtils.generate_random_bytes(32)
        payload = {
            "createEphemeral": True,
            "elevate": True,
            "elevationMethod": "group",
            "elevationString": "arn:aws",
            "baseDistinguishedName": ""
        }
        link = encrypted_link(payload, key, path="jit_settings")

        data = link.get_jit_settings_data(key)
        self.assertIsNotNone(data)
        self.assertEqual(payload, data)

    def test_losslessness(self):
        """Unknown link keys and payload fields are preserved in raw and get_link_data."""

        payload = {"is_admin": True, "futureField": {"nested": [1, 2, 3]}}
        link_dict = {
            "recordUid": "RU",
            "data": utils.bytes_to_base64(json.dumps(payload).encode()),
            "path": None,
            "futureLinkKey": "kept"
        }
        link = KeeperRecordLink(link_dict)

        self.assertEqual(link_dict, link.raw, "raw keeps the original dict untouched")
        self.assertEqual("kept", link.raw.get("futureLinkKey"))

        data = link.get_link_data()
        self.assertEqual({"nested": [1, 2, 3]}, data.get("futureField"),
                         "unknown payload fields pass through get_link_data")

    def test_top_level_wins_over_allowed_settings(self):
        """A top-level boolean takes precedence over the allowedSettings fallback."""

        link = plain_link({
            "rotation": False,
            "allowedSettings": {"rotation": True}
        })
        self.assertFalse(link.allows_rotation(), "top-level value wins")

        only_nested = plain_link({"allowedSettings": {"rotation": True}})
        self.assertTrue(only_nested.allows_rotation(), "fallback applies when top level is absent")

    def test_ciphertext_with_json_like_first_byte(self):
        """Ciphertext coincidentally starting with "{" or "[" still decrypts.

        AES-GCM output starts with the random IV, so ~2/256 of encrypted links
        begin with a JSON marker. The plain-JSON fast path must fall through to
        decryption when its parse fails, instead of dropping the data.
        """

        key = CryptoUtils.generate_random_bytes(32)
        payload = {"createEphemeral": True, "elevate": True}

        for marker in (b"{", b"["):
            iv = marker + os.urandom(11)
            ciphertext = CryptoUtils.encrypt_aes(json.dumps(payload).encode(), key, iv=iv)
            link = KeeperRecordLink({
                "recordUid": "RU",
                "data": utils.bytes_to_base64(ciphertext),
                "path": "jit_settings"
            })

            decoded = link.get_decoded_data()
            self.assertEqual(marker.decode(), decoded[0],
                             "fixture must start with the JSON marker")

            self.assertEqual(payload, link.get_link_data(key),
                             "falls through to decryption despite the JSON-like first byte")
            self.assertEqual(payload, link.get_jit_settings_data(key),
                             "settings accessors benefit from the fall-through")
            self.assertEqual(payload, link.get_settings_for_path("jit_settings", key))
            self.assertIsNone(link.get_link_data(None),
                              "still None without a key")

        # The plain-JSON fast path is unaffected.
        self.assertEqual({"a": 1}, plain_link({"a": 1}).get_link_data())


if __name__ == '__main__':
    unittest.main()
